src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -620,13 +629,29 @@ private static Expression GetValueFromProperty(Expression sourceExpression, stri
             return Expression.Convert(indexExpression, typeof(string));
         }
 
-        private static Expression BindParameterFromService(ParameterInfo parameter)
+        private static Expression BindParameterFromService(ParameterInfo parameter, FactoryContext factoryContext)
         {
             var isOptional = IsOptionalParameter(parameter);
 
-            return isOptional
-                ? Expression.Call(GetServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr)
-                : Expression.Call(GetRequiredServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr);
+            if (isOptional)
+            {
+                return Expression.Call(GetServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr);
+            }
+            if (factoryContext.ServiceProviderIsService is IServiceProviderIsService serviceProviderIsService)
+            {
+                if (serviceProviderIsService.IsService(parameter.ParameterType))
+                {
+                    return Expression.Call(GetRequiredServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr);
+                }
+            }
+            else
+            {
+                // If the DI container does not implement IServiceProviderIsService then we can't error early about the service not being in the container
+                return Expression.Call(GetRequiredServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr);
+            }
+
+            var errorMessage = BuildErrorMessageForMissingService(factoryContext);
```

Comment:
Since this error message references includes all the `TrackedParameters`, we should only throw after all the parameters have been evaluated to give a more complete error message.